// @layerzerolabs/lz-definitions/dist/index.cjs
'use strict';

var invariant = require('tiny-invariant');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var invariant__default = /*#__PURE__*/_interopDefault(invariant);

// src/enums/chain.ts
var Chain = /* @__PURE__ */ ((Chain2) => {
  Chain2["ETHEREUM"] = "ethereum";
  Chain2["BSC"] = "bsc";
  Chain2["AVALANCHE"] = "avalanche";
  Chain2["POLYGON"] = "polygon";
  Chain2["ZKPOLYGON"] = "zkpolygon";
  Chain2["ARBITRUM"] = "arbitrum";
  Chain2["OPTIMISM"] = "optimism";
  Chain2["FANTOM"] = "fantom";
  Chain2["SWIMMER"] = "swimmer";
  Chain2["DFK"] = "dfk";
  Chain2["HARMONY"] = "harmony";
  Chain2["ARCANA"] = "arcana";
  Chain2["DEXALOT"] = "dexalot";
  Chain2["CASTLECRUSH"] = "castlecrush";
  Chain2["GNOSIS"] = "gnosis";
  Chain2["CELO"] = "celo";
  Chain2["MOONBEAM"] = "moonbeam";
  Chain2["BOBA"] = "boba";
  Chain2["PORTAL"] = "portal";
  Chain2["INTAIN"] = "intain";
  Chain2["FUSE"] = "fuse";
  Chain2["ASTAR"] = "astar";
  Chain2["ZKSYNC"] = "zksync";
  Chain2["SHRAPNEL"] = "shrapnel";
  Chain2["DOS"] = "dos";
  Chain2["APTOS"] = "aptos";
  Chain2["KLAYTN"] = "klaytn";
  Chain2["METIS"] = "metis";
  Chain2["COREDAO"] = "coredao";
  Chain2["OKX"] = "okx";
  Chain2["METER"] = "meter";
  Chain2["GOERLI"] = "goerli";
  Chain2["SOLANA"] = "solana";
  Chain2["BASE"] = "base";
  Chain2["ZKCONSENSYS"] = "zkconsensys";
  Chain2["SEPOLIA"] = "sepolia";
  Chain2["MOONRIVER"] = "moonriver";
  Chain2["AAVEGOTCHI"] = "aavegotchi";
  Chain2["SCROLL"] = "scroll";
  Chain2["CATHAY"] = "cathay";
  Chain2["KAVA"] = "kava";
  Chain2["TENET"] = "tenet";
  Chain2["CANTO"] = "canto";
  Chain2["ORDERLY"] = "orderly";
  Chain2["NOVA"] = "nova";
  Chain2["BLOCKGEN"] = "blockgen";
  Chain2["MERITCIRCLE"] = "meritcircle";
  Chain2["MANTLE"] = "mantle";
  Chain2["HUBBLE"] = "hubble";
  Chain2["ZORA"] = "zora";
  Chain2["TOMO"] = "tomo";
  Chain2["LOOT"] = "loot";
  Chain2["TELOS"] = "telos";
  Chain2["AURORA"] = "aurora";
  Chain2["OPBNB"] = "opbnb";
  Chain2["SHIMMER"] = "shimmer";
  Chain2["MONAD"] = "monad";
  Chain2["LIF3"] = "lif3";
  Chain2["SPRUCE"] = "spruce";
  Chain2["PGJJTK"] = "pgjjtk";
  Chain2["ODA"] = "oda";
  Chain2["KIWI"] = "kiwi";
  Chain2["CONFLUX"] = "conflux";
  Chain2["EON"] = "eon";
  Chain2["XPLA"] = "xpla";
  Chain2["HOLESKY"] = "holesky";
  Chain2["INJECTIVE"] = "injective";
  Chain2["IDEX"] = "idex";
  Chain2["ZKATANA"] = "zkatana";
  Chain2["MANTA"] = "manta";
  Chain2["FRAME"] = "frame";
  Chain2["PGN"] = "pgn";
  Chain2["POLYGONCDK"] = "polygoncdk";
  Chain2["ARBSEP"] = "arbsep";
  Chain2["OPTSEP"] = "optsep";
  return Chain2;
})(Chain || {});

// src/enums/chain-type.ts
var ChainType = /* @__PURE__ */ ((ChainType2) => {
  ChainType2["UNKNOWN"] = "unknown";
  ChainType2["EVM"] = "evm";
  ChainType2["APTOS"] = "aptos";
  ChainType2["SOLANA"] = "solana";
  return ChainType2;
})(ChainType || {});

// src/enums/endpoint-id.ts
var MainnetEndpointId = /* @__PURE__ */ ((MainnetEndpointId2) => {
  MainnetEndpointId2[MainnetEndpointId2["ETHEREUM_MAINNET"] = 101] = "ETHEREUM_MAINNET";
  MainnetEndpointId2[MainnetEndpointId2["BSC_MAINNET"] = 102] = "BSC_MAINNET";
  MainnetEndpointId2[MainnetEndpointId2["AVALANCHE_MAINNET"] = 106] = "AVALANCHE_MAINNET";
  MainnetEndpointId2[MainnetEndpointId2["POLYGON_MAINNET"] = 109] = "POLYGON_MAINNET";
  MainnetEndpointId2[MainnetEndpointId2["ARBITRUM_MAINNET"] = 110] = "ARBITRUM_MAINNET";
  MainnetEndpointId2[MainnetEndpointId2["OPTIMISM_MAINNET"] = 111] = "OPTIMISM_MAINNET";
  MainnetEndpointId2[MainnetEndpointId2["FANTOM_MAINNET"] = 112] = "FANTOM_MAINNET";
  MainnetEndpointId2[MainnetEndpointId2["SWIMMER_MAINNET"] = 114] = "SWIMMER_MAINNET";
  MainnetEndpointId2[MainnetEndpointId2["DFK_MAINNET"] = 115] = "DFK_MAINNET";
  MainnetEndpointId2[MainnetEndpointId2["HARMONY_MAINNET"] = 116] = "HARMONY_MAINNET";
  MainnetEndpointId2[MainnetEndpointId2["DEXALOT_MAINNET"] = 118] = "DEXALOT_MAINNET";
  MainnetEndpointId2[MainnetEndpointId2["CELO_MAINNET"] = 125] = "CELO_MAINNET";
  MainnetEndpointId2[MainnetEndpointId2["MOONBEAM_MAINNET"] = 126] = "MOONBEAM_MAINNET";
  MainnetEndpointId2[MainnetEndpointId2["INTAIN_MAINNET"] = 152] = "INTAIN_MAINNET";
  MainnetEndpointId2[MainnetEndpointId2["FUSE_MAINNET"] = 138] = "FUSE_MAINNET";
  MainnetEndpointId2[MainnetEndpointId2["GNOSIS_MAINNET"] = 145] = "GNOSIS_MAINNET";
  MainnetEndpointId2[MainnetEndpointId2["APTOS_MAINNET"] = 108] = "APTOS_MAINNET";
  MainnetEndpointId2[MainnetEndpointId2["KLAYTN_MAINNET"] = 150] = "KLAYTN_MAINNET";
  MainnetEndpointId2[MainnetEndpointId2["METIS_MAINNET"] = 151] = "METIS_MAINNET";
  MainnetEndpointId2[MainnetEndpointId2["COREDAO_MAINNET"] = 153] = "COREDAO_MAINNET";
  MainnetEndpointId2[MainnetEndpointId2["GOERLI_MAINNET"] = 154] = "GOERLI_MAINNET";
  MainnetEndpointId2[MainnetEndpointId2["OKX_MAINNET"] = 155] = "OKX_MAINNET";
  MainnetEndpointId2[MainnetEndpointId2["CANTO_MAINNET"] = 159] = "CANTO_MAINNET";
  MainnetEndpointId2[MainnetEndpointId2["ZKSYNC_MAINNET"] = 165] = "ZKSYNC_MAINNET";
  MainnetEndpointId2[MainnetEndpointId2["SHRAPNEL_MAINNET"] = 148] = "SHRAPNEL_MAINNET";
  MainnetEndpointId2[MainnetEndpointId2["DOS_MAINNET"] = 149] = "DOS_MAINNET";
  MainnetEndpointId2[MainnetEndpointId2["ZKPOLYGON_MAINNET"] = 158] = "ZKPOLYGON_MAINNET";
  MainnetEndpointId2[MainnetEndpointId2["SEPOLIA_MAINNET"] = 161] = "SEPOLIA_MAINNET";
  MainnetEndpointId2[MainnetEndpointId2["MOONRIVER_MAINNET"] = 167] = "MOONRIVER_MAINNET";
  MainnetEndpointId2[MainnetEndpointId2["SOLANA_MAINNET"] = 168] = "SOLANA_MAINNET";
  MainnetEndpointId2[MainnetEndpointId2["TENET_MAINNET"] = 173] = "TENET_MAINNET";
  MainnetEndpointId2[MainnetEndpointId2["NOVA_MAINNET"] = 175] = "NOVA_MAINNET";
  MainnetEndpointId2[MainnetEndpointId2["METER_MAINNET"] = 176] = "METER_MAINNET";
  MainnetEndpointId2[MainnetEndpointId2["KAVA_MAINNET"] = 177] = "KAVA_MAINNET";
  MainnetEndpointId2[MainnetEndpointId2["MANTLE_MAINNET"] = 181] = "MANTLE_MAINNET";
  MainnetEndpointId2[MainnetEndpointId2["HUBBLE_MAINNET"] = 182] = "HUBBLE_MAINNET";
  MainnetEndpointId2[MainnetEndpointId2["ZKCONSENSYS_MAINNET"] = 183] = "ZKCONSENSYS_MAINNET";
  MainnetEndpointId2[MainnetEndpointId2["BASE_MAINNET"] = 184] = "BASE_MAINNET";
  MainnetEndpointId2[MainnetEndpointId2["ZORA_MAINNET"] = 195] = "ZORA_MAINNET";
  MainnetEndpointId2[MainnetEndpointId2["TOMO_MAINNET"] = 196] = "TOMO_MAINNET";
  MainnetEndpointId2[MainnetEndpointId2["LOOT_MAINNET"] = 197] = "LOOT_MAINNET";
  MainnetEndpointId2[MainnetEndpointId2["MERITCIRCLE_MAINNET"] = 198] = "MERITCIRCLE_MAINNET";
  MainnetEndpointId2[MainnetEndpointId2["TELOS_MAINNET"] = 199] = "TELOS_MAINNET";
  MainnetEndpointId2[MainnetEndpointId2["OPBNB_MAINNET"] = 202] = "OPBNB_MAINNET";
  MainnetEndpointId2[MainnetEndpointId2["LIF3_MAINNET"] = 205] = "LIF3_MAINNET";
  MainnetEndpointId2[MainnetEndpointId2["ASTAR_MAINNET"] = 210] = "ASTAR_MAINNET";
  MainnetEndpointId2[MainnetEndpointId2["AURORA_MAINNET"] = 211] = "AURORA_MAINNET";
  MainnetEndpointId2[MainnetEndpointId2["CONFLUX_MAINNET"] = 212] = "CONFLUX_MAINNET";
  MainnetEndpointId2[MainnetEndpointId2["ORDERLY_MAINNET"] = 213] = "ORDERLY_MAINNET";
  MainnetEndpointId2[MainnetEndpointId2["SCROLL_MAINNET"] = 214] = "SCROLL_MAINNET";
  MainnetEndpointId2[MainnetEndpointId2["EON_MAINNET"] = 215] = "EON_MAINNET";
  MainnetEndpointId2[MainnetEndpointId2["XPLA_MAINNET"] = 216] = "XPLA_MAINNET";
  MainnetEndpointId2[MainnetEndpointId2["MANTA_MAINNET"] = 217] = "MANTA_MAINNET";
  MainnetEndpointId2[MainnetEndpointId2["PGN_MAINNET"] = 218] = "PGN_MAINNET";
  MainnetEndpointId2[MainnetEndpointId2["SHIMMER_MAINNET"] = 230] = "SHIMMER_MAINNET";
  return MainnetEndpointId2;
})(MainnetEndpointId || {});
var TestnetEndpointId = /* @__PURE__ */ ((TestnetEndpointId2) => {
  TestnetEndpointId2[TestnetEndpointId2["BSC_TESTNET"] = 10102] = "BSC_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["AVALANCHE_TESTNET"] = 10106] = "AVALANCHE_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["POLYGON_TESTNET"] = 10109] = "POLYGON_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["FANTOM_TESTNET"] = 10112] = "FANTOM_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["SWIMMER_TESTNET"] = 10130] = "SWIMMER_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["DFK_TESTNET"] = 10115] = "DFK_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["HARMONY_TESTNET"] = 10133] = "HARMONY_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["DEXALOT_TESTNET"] = 10118] = "DEXALOT_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["CELO_TESTNET"] = 10125] = "CELO_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["MOONBEAM_TESTNET"] = 10126] = "MOONBEAM_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["INTAIN_TESTNET"] = 10152] = "INTAIN_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["ETHEREUM_TESTNET"] = 10121] = "ETHEREUM_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["ARBITRUM_TESTNET"] = 10143] = "ARBITRUM_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["OPTIMISM_TESTNET"] = 10132] = "OPTIMISM_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["FUSE_TESTNET"] = 10138] = "FUSE_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["GNOSIS_TESTNET"] = 10145] = "GNOSIS_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["ZKSYNC_TESTNET"] = 10165] = "ZKSYNC_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["SHRAPNEL_TESTNET"] = 10164] = "SHRAPNEL_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["DOS_TESTNET"] = 10162] = "DOS_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["APTOS_TESTNET"] = 10108] = "APTOS_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["KLAYTN_TESTNET"] = 10150] = "KLAYTN_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["METIS_TESTNET"] = 10151] = "METIS_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["COREDAO_TESTNET"] = 10153] = "COREDAO_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["OKX_TESTNET"] = 10155] = "OKX_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["METER_TESTNET"] = 10156] = "METER_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["ZKCONSENSYS_TESTNET"] = 10157] = "ZKCONSENSYS_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["ZKPOLYGON_TESTNET"] = 10158] = "ZKPOLYGON_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["CANTO_TESTNET"] = 10159] = "CANTO_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["BASE_TESTNET"] = 10160] = "BASE_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["SEPOLIA_TESTNET"] = 10161] = "SEPOLIA_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["SOLANA_TESTNET"] = 10168] = "SOLANA_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["CATHAY_TESTNET"] = 10171] = "CATHAY_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["KAVA_TESTNET"] = 10172] = "KAVA_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["TENET_TESTNET"] = 10173] = "TENET_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["BLOCKGEN_TESTNET"] = 10177] = "BLOCKGEN_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["MERITCIRCLE_TESTNET"] = 10178] = "MERITCIRCLE_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["MANTLE_TESTNET"] = 10181] = "MANTLE_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["HUBBLE_TESTNET"] = 10182] = "HUBBLE_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["AAVEGOTCHI_TESTNET"] = 10191] = "AAVEGOTCHI_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["ZORA_TESTNET"] = 10195] = "ZORA_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["TOMO_TESTNET"] = 10196] = "TOMO_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["LOOT_TESTNET"] = 10197] = "LOOT_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["TELOS_TESTNET"] = 10199] = "TELOS_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["ORDERLY_TESTNET"] = 10200] = "ORDERLY_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["AURORA_TESTNET"] = 10201] = "AURORA_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["OPBNB_TESTNET"] = 10202] = "OPBNB_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["MONAD_TESTNET"] = 10204] = "MONAD_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["LIF3_TESTNET"] = 10205] = "LIF3_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["SPRUCE_TESTNET"] = 10206] = "SPRUCE_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["PGJJTK_TESTNET"] = 10207] = "PGJJTK_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["ODA_TESTNET"] = 10208] = "ODA_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["KIWI_TESTNET"] = 10209] = "KIWI_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["ASTAR_TESTNET"] = 10210] = "ASTAR_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["CONFLUX_TESTNET"] = 10211] = "CONFLUX_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["SCROLL_TESTNET"] = 10214] = "SCROLL_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["EON_TESTNET"] = 10215] = "EON_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["XPLA_TESTNET"] = 10216] = "XPLA_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["HOLESKY_TESTNET"] = 10217] = "HOLESKY_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["INJECTIVE_TESTNET"] = 10218] = "INJECTIVE_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["IDEX_TESTNET"] = 10219] = "IDEX_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["ZKATANA_TESTNET"] = 10220] = "ZKATANA_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["MANTA_TESTNET"] = 10221] = "MANTA_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["FRAME_TESTNET"] = 10222] = "FRAME_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["PGN_TESTNET"] = 10223] = "PGN_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["POLYGONCDK_TESTNET"] = 10224] = "POLYGONCDK_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["SHIMMER_TESTNET"] = 10230] = "SHIMMER_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["ARBSEP_TESTNET"] = 10231] = "ARBSEP_TESTNET";
  TestnetEndpointId2[TestnetEndpointId2["OPTSEP_TESTNET"] = 10232] = "OPTSEP_TESTNET";
  return TestnetEndpointId2;
})(TestnetEndpointId || {});
var SandboxEndpointId = /* @__PURE__ */ ((SandboxEndpointId2) => {
  SandboxEndpointId2[SandboxEndpointId2["BSC_SANDBOX"] = 20102] = "BSC_SANDBOX";
  SandboxEndpointId2[SandboxEndpointId2["AVALANCHE_SANDBOX"] = 20106] = "AVALANCHE_SANDBOX";
  SandboxEndpointId2[SandboxEndpointId2["POLYGON_SANDBOX"] = 20109] = "POLYGON_SANDBOX";
  SandboxEndpointId2[SandboxEndpointId2["ETHEREUM_SANDBOX"] = 20121] = "ETHEREUM_SANDBOX";
  SandboxEndpointId2[SandboxEndpointId2["APTOS_SANDBOX"] = 20008] = "APTOS_SANDBOX";
  return SandboxEndpointId2;
})(SandboxEndpointId || {});
var MainnetV2EndpointId = /* @__PURE__ */ ((MainnetV2EndpointId2) => {
  MainnetV2EndpointId2[MainnetV2EndpointId2["ETHEREUM_V2_MAINNET"] = 30101] = "ETHEREUM_V2_MAINNET";
  MainnetV2EndpointId2[MainnetV2EndpointId2["BSC_V2_MAINNET"] = 30102] = "BSC_V2_MAINNET";
  MainnetV2EndpointId2[MainnetV2EndpointId2["AVALANCHE_V2_MAINNET"] = 30106] = "AVALANCHE_V2_MAINNET";
  MainnetV2EndpointId2[MainnetV2EndpointId2["POLYGON_V2_MAINNET"] = 30109] = "POLYGON_V2_MAINNET";
  MainnetV2EndpointId2[MainnetV2EndpointId2["ARBITRUM_V2_MAINNET"] = 30110] = "ARBITRUM_V2_MAINNET";
  MainnetV2EndpointId2[MainnetV2EndpointId2["OPTIMISM_V2_MAINNET"] = 30111] = "OPTIMISM_V2_MAINNET";
  MainnetV2EndpointId2[MainnetV2EndpointId2["FANTOM_V2_MAINNET"] = 30112] = "FANTOM_V2_MAINNET";
  MainnetV2EndpointId2[MainnetV2EndpointId2["SWIMMER_V2_MAINNET"] = 30114] = "SWIMMER_V2_MAINNET";
  MainnetV2EndpointId2[MainnetV2EndpointId2["DFK_V2_MAINNET"] = 30115] = "DFK_V2_MAINNET";
  MainnetV2EndpointId2[MainnetV2EndpointId2["HARMONY_V2_MAINNET"] = 30116] = "HARMONY_V2_MAINNET";
  MainnetV2EndpointId2[MainnetV2EndpointId2["DEXALOT_V2_MAINNET"] = 30118] = "DEXALOT_V2_MAINNET";
  MainnetV2EndpointId2[MainnetV2EndpointId2["CELO_V2_MAINNET"] = 30125] = "CELO_V2_MAINNET";
  MainnetV2EndpointId2[MainnetV2EndpointId2["MOONBEAM_V2_MAINNET"] = 30126] = "MOONBEAM_V2_MAINNET";
  MainnetV2EndpointId2[MainnetV2EndpointId2["INTAIN_V2_MAINNET"] = 30152] = "INTAIN_V2_MAINNET";
  MainnetV2EndpointId2[MainnetV2EndpointId2["FUSE_V2_MAINNET"] = 30138] = "FUSE_V2_MAINNET";
  MainnetV2EndpointId2[MainnetV2EndpointId2["GNOSIS_V2_MAINNET"] = 30145] = "GNOSIS_V2_MAINNET";
  MainnetV2EndpointId2[MainnetV2EndpointId2["APTOS_V2_MAINNET"] = 30108] = "APTOS_V2_MAINNET";
  MainnetV2EndpointId2[MainnetV2EndpointId2["KLAYTN_V2_MAINNET"] = 30150] = "KLAYTN_V2_MAINNET";
  MainnetV2EndpointId2[MainnetV2EndpointId2["METIS_V2_MAINNET"] = 30151] = "METIS_V2_MAINNET";
  MainnetV2EndpointId2[MainnetV2EndpointId2["COREDAO_V2_MAINNET"] = 30153] = "COREDAO_V2_MAINNET";
  MainnetV2EndpointId2[MainnetV2EndpointId2["GOERLI_V2_MAINNET"] = 30154] = "GOERLI_V2_MAINNET";
  MainnetV2EndpointId2[MainnetV2EndpointId2["OKX_V2_MAINNET"] = 30155] = "OKX_V2_MAINNET";
  MainnetV2EndpointId2[MainnetV2EndpointId2["CANTO_V2_MAINNET"] = 30159] = "CANTO_V2_MAINNET";
  MainnetV2EndpointId2[MainnetV2EndpointId2["ZKSYNC_V2_MAINNET"] = 30165] = "ZKSYNC_V2_MAINNET";
  MainnetV2EndpointId2[MainnetV2EndpointId2["SHRAPNEL_V2_MAINNET"] = 30148] = "SHRAPNEL_V2_MAINNET";
  MainnetV2EndpointId2[MainnetV2EndpointId2["DOS_V2_MAINNET"] = 30149] = "DOS_V2_MAINNET";
  MainnetV2EndpointId2[MainnetV2EndpointId2["ZKPOLYGON_V2_MAINNET"] = 30158] = "ZKPOLYGON_V2_MAINNET";
  MainnetV2EndpointId2[MainnetV2EndpointId2["SEPOLIA_V2_MAINNET"] = 30161] = "SEPOLIA_V2_MAINNET";
  MainnetV2EndpointId2[MainnetV2EndpointId2["MOONRIVER_V2_MAINNET"] = 30167] = "MOONRIVER_V2_MAINNET";
  MainnetV2EndpointId2[MainnetV2EndpointId2["SOLANA_V2_MAINNET"] = 30168] = "SOLANA_V2_MAINNET";
  MainnetV2EndpointId2[MainnetV2EndpointId2["TENET_V2_MAINNET"] = 30173] = "TENET_V2_MAINNET";
  MainnetV2EndpointId2[MainnetV2EndpointId2["NOVA_V2_MAINNET"] = 30175] = "NOVA_V2_MAINNET";
  MainnetV2EndpointId2[MainnetV2EndpointId2["METER_V2_MAINNET"] = 30176] = "METER_V2_MAINNET";
  MainnetV2EndpointId2[MainnetV2EndpointId2["KAVA_V2_MAINNET"] = 30177] = "KAVA_V2_MAINNET";
  MainnetV2EndpointId2[MainnetV2EndpointId2["MANTLE_V2_MAINNET"] = 30181] = "MANTLE_V2_MAINNET";
  MainnetV2EndpointId2[MainnetV2EndpointId2["HUBBLE_V2_MAINNET"] = 30182] = "HUBBLE_V2_MAINNET";
  MainnetV2EndpointId2[MainnetV2EndpointId2["ZKCONSENSYS_V2_MAINNET"] = 30183] = "ZKCONSENSYS_V2_MAINNET";
  MainnetV2EndpointId2[MainnetV2EndpointId2["BASE_V2_MAINNET"] = 30184] = "BASE_V2_MAINNET";
  MainnetV2EndpointId2[MainnetV2EndpointId2["ZORA_V2_MAINNET"] = 30195] = "ZORA_V2_MAINNET";
  MainnetV2EndpointId2[MainnetV2EndpointId2["TOMO_V2_MAINNET"] = 30196] = "TOMO_V2_MAINNET";
  MainnetV2EndpointId2[MainnetV2EndpointId2["LOOT_V2_MAINNET"] = 30197] = "LOOT_V2_MAINNET";
  MainnetV2EndpointId2[MainnetV2EndpointId2["MERITCIRCLE_V2_MAINNET"] = 30198] = "MERITCIRCLE_V2_MAINNET";
  MainnetV2EndpointId2[MainnetV2EndpointId2["TELOS_V2_MAINNET"] = 30199] = "TELOS_V2_MAINNET";
  MainnetV2EndpointId2[MainnetV2EndpointId2["OPBNB_V2_MAINNET"] = 30202] = "OPBNB_V2_MAINNET";
  MainnetV2EndpointId2[MainnetV2EndpointId2["LIF3_V2_MAINNET"] = 30205] = "LIF3_V2_MAINNET";
  MainnetV2EndpointId2[MainnetV2EndpointId2["ASTAR_V2_MAINNET"] = 30210] = "ASTAR_V2_MAINNET";
  MainnetV2EndpointId2[MainnetV2EndpointId2["AURORA_V2_MAINNET"] = 30211] = "AURORA_V2_MAINNET";
  MainnetV2EndpointId2[MainnetV2EndpointId2["CONFLUX_V2_MAINNET"] = 30212] = "CONFLUX_V2_MAINNET";
  MainnetV2EndpointId2[MainnetV2EndpointId2["ORDERLY_V2_MAINNET"] = 30213] = "ORDERLY_V2_MAINNET";
  MainnetV2EndpointId2[MainnetV2EndpointId2["SCROLL_V2_MAINNET"] = 30214] = "SCROLL_V2_MAINNET";
  MainnetV2EndpointId2[MainnetV2EndpointId2["EON_V2_MAINNET"] = 30215] = "EON_V2_MAINNET";
  MainnetV2EndpointId2[MainnetV2EndpointId2["XPLA_V2_MAINNET"] = 30216] = "XPLA_V2_MAINNET";
  MainnetV2EndpointId2[MainnetV2EndpointId2["MANTA_V2_MAINNET"] = 30217] = "MANTA_V2_MAINNET";
  MainnetV2EndpointId2[MainnetV2EndpointId2["PGN_V2_MAINNET"] = 30218] = "PGN_V2_MAINNET";
  MainnetV2EndpointId2[MainnetV2EndpointId2["SHIMMER_V2_MAINNET"] = 30230] = "SHIMMER_V2_MAINNET";
  return MainnetV2EndpointId2;
})(MainnetV2EndpointId || {});
var TestnetV2EndpointId = /* @__PURE__ */ ((TestnetV2EndpointId2) => {
  TestnetV2EndpointId2[TestnetV2EndpointId2["AVALANCHE_V2_TESTNET"] = 40106] = "AVALANCHE_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["POLYGON_V2_TESTNET"] = 40109] = "POLYGON_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["ETHEREUM_V2_TESTNET"] = 40121] = "ETHEREUM_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["OPTIMISM_V2_TESTNET"] = 40132] = "OPTIMISM_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["ARBITRUM_V2_TESTNET"] = 40143] = "ARBITRUM_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["SOLANA_V2_TESTNET"] = 40168] = "SOLANA_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["BSC_V2_TESTNET"] = 40102] = "BSC_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["FANTOM_V2_TESTNET"] = 40112] = "FANTOM_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["DFK_V2_TESTNET"] = 40115] = "DFK_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["HARMONY_V2_TESTNET"] = 40133] = "HARMONY_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["DEXALOT_V2_TESTNET"] = 40118] = "DEXALOT_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["CELO_V2_TESTNET"] = 40125] = "CELO_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["MOONBEAM_V2_TESTNET"] = 40126] = "MOONBEAM_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["FUSE_V2_TESTNET"] = 40138] = "FUSE_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["GNOSIS_V2_TESTNET"] = 40145] = "GNOSIS_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["ZKSYNC_V2_TESTNET"] = 40165] = "ZKSYNC_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["SHRAPNEL_V2_TESTNET"] = 40164] = "SHRAPNEL_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["DOS_V2_TESTNET"] = 40162] = "DOS_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["KLAYTN_V2_TESTNET"] = 40150] = "KLAYTN_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["METIS_V2_TESTNET"] = 40151] = "METIS_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["COREDAO_V2_TESTNET"] = 40153] = "COREDAO_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["OKX_V2_TESTNET"] = 40155] = "OKX_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["METER_V2_TESTNET"] = 40156] = "METER_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["ZKCONSENSYS_V2_TESTNET"] = 40157] = "ZKCONSENSYS_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["ZKPOLYGON_V2_TESTNET"] = 40158] = "ZKPOLYGON_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["CANTO_V2_TESTNET"] = 40159] = "CANTO_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["BASE_V2_TESTNET"] = 40160] = "BASE_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["SEPOLIA_V2_TESTNET"] = 40161] = "SEPOLIA_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["AAVEGOTCHI_V2_TESTNET"] = 40179] = "AAVEGOTCHI_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["SCROLL_V2_TESTNET"] = 40170] = "SCROLL_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["CATHAY_V2_TESTNET"] = 40171] = "CATHAY_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["KAVA_V2_TESTNET"] = 40172] = "KAVA_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["TENET_V2_TESTNET"] = 40173] = "TENET_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["BLOCKGEN_V2_TESTNET"] = 40177] = "BLOCKGEN_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["MERITCIRCLE_V2_TESTNET"] = 40178] = "MERITCIRCLE_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["MANTLE_V2_TESTNET"] = 40181] = "MANTLE_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["HUBBLE_V2_TESTNET"] = 40182] = "HUBBLE_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["ZORA_V2_TESTNET"] = 40195] = "ZORA_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["TOMO_V2_TESTNET"] = 40196] = "TOMO_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["LOOT_V2_TESTNET"] = 40197] = "LOOT_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["TELOS_V2_TESTNET"] = 40199] = "TELOS_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["ORDERLY_V2_TESTNET"] = 40200] = "ORDERLY_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["AURORA_V2_TESTNET"] = 40201] = "AURORA_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["OPBNB_V2_TESTNET"] = 40202] = "OPBNB_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["SHIMMER_V2_TESTNET"] = 40203] = "SHIMMER_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["MONAD_V2_TESTNET"] = 40204] = "MONAD_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["LIF3_V2_TESTNET"] = 40205] = "LIF3_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["SPRUCE_V2_TESTNET"] = 40206] = "SPRUCE_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["PGJJTK_V2_TESTNET"] = 40207] = "PGJJTK_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["ODA_V2_TESTNET"] = 40208] = "ODA_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["KIWI_V2_TESTNET"] = 40209] = "KIWI_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["ASTAR_V2_TESTNET"] = 40210] = "ASTAR_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["CONFLUX_V2_TESTNET"] = 40211] = "CONFLUX_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["EON_V2_TESTNET"] = 40215] = "EON_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["XPLA_V2_TESTNET"] = 40216] = "XPLA_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["HOLESKY_V2_TESTNET"] = 40217] = "HOLESKY_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["INJECTIVE_V2_TESTNET"] = 40218] = "INJECTIVE_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["IDEX_V2_TESTNET"] = 40219] = "IDEX_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["ZKATANA_V2_TESTNET"] = 40220] = "ZKATANA_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["MANTA_V2_TESTNET"] = 40221] = "MANTA_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["FRAME_V2_TESTNET"] = 40222] = "FRAME_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["PGN_V2_TESTNET"] = 40223] = "PGN_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["POLYGONCDK_V2_TESTNET"] = 40224] = "POLYGONCDK_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["ARBSEP_V2_TESTNET"] = 40231] = "ARBSEP_V2_TESTNET";
  TestnetV2EndpointId2[TestnetV2EndpointId2["OPTSEP_V2_TESTNET"] = 40232] = "OPTSEP_V2_TESTNET";
  return TestnetV2EndpointId2;
})(TestnetV2EndpointId || {});
var SandboxV2EndpointId = /* @__PURE__ */ ((SandboxV2EndpointId2) => {
  SandboxV2EndpointId2[SandboxV2EndpointId2["ETHEREUM_V2_SANDBOX"] = 50121] = "ETHEREUM_V2_SANDBOX";
  SandboxV2EndpointId2[SandboxV2EndpointId2["POLYGON_V2_SANDBOX"] = 50109] = "POLYGON_V2_SANDBOX";
  SandboxV2EndpointId2[SandboxV2EndpointId2["SOLANA_V2_SANDBOX"] = 50168] = "SOLANA_V2_SANDBOX";
  SandboxV2EndpointId2[SandboxV2EndpointId2["BSC_V2_SANDBOX"] = 50102] = "BSC_V2_SANDBOX";
  SandboxV2EndpointId2[SandboxV2EndpointId2["AVALANCHE_V2_SANDBOX"] = 50106] = "AVALANCHE_V2_SANDBOX";
  return SandboxV2EndpointId2;
})(SandboxV2EndpointId || {});
var EndpointId = {
  ...MainnetEndpointId,
  ...TestnetEndpointId,
  ...SandboxEndpointId,
  ...MainnetV2EndpointId,
  ...TestnetV2EndpointId,
  ...SandboxV2EndpointId
};

// src/enums/environment.ts
var Environment = /* @__PURE__ */ ((Environment2) => {
  Environment2["LOCAL"] = "local";
  Environment2["DEVNET"] = "devnet";
  Environment2["TESTNET"] = "testnet";
  Environment2["MAINNET"] = "mainnet";
  return Environment2;
})(Environment || {});

// src/enums/stage.ts
var Stage = /* @__PURE__ */ ((Stage2) => {
  Stage2["MAINNET"] = "mainnet";
  Stage2["TESTNET"] = "testnet";
  Stage2["SANDBOX"] = "sandbox";
  return Stage2;
})(Stage || {});

// src/enums/endpoint-version.ts
var EndpointVersion = /* @__PURE__ */ ((EndpointVersion2) => {
  EndpointVersion2["V1"] = "v1";
  EndpointVersion2["V2"] = "v2";
  return EndpointVersion2;
})(EndpointVersion || {});

// src/enums/message-lib.ts
var MessageLibVersionV2 = /* @__PURE__ */ ((MessageLibVersionV22) => {
  MessageLibVersionV22["SIMPLE"] = "0.0.2";
  MessageLibVersionV22["ULTRA_LIGHT_NODE_302"] = "3.0.2";
  return MessageLibVersionV22;
})(MessageLibVersionV2 || {});
var MessageLibVersionV1 = /* @__PURE__ */ ((MessageLibVersionV12) => {
  MessageLibVersionV12["ULTRA_LIGHT_NODE_301"] = "3.0.1";
  MessageLibVersionV12["ULTRA_LIGHT_NODE_201"] = "2.0.1";
  return MessageLibVersionV12;
})(MessageLibVersionV1 || {});

// src/enums/chain-key.ts
var ChainKey = /* @__PURE__ */ ((ChainKey3) => {
  ChainKey3["ETHEREUM"] = "ethereum";
  ChainKey3["GOERLI"] = "goerli";
  ChainKey3["SEPOLIA"] = "sepolia";
  ChainKey3["BSC"] = "bsc";
  ChainKey3["BSC_TESTNET"] = "bsc-testnet";
  ChainKey3["AVALANCHE"] = "avalanche";
  ChainKey3["FUJI"] = "fuji";
  ChainKey3["POLYGON"] = "polygon";
  ChainKey3["MUMBAI"] = "mumbai";
  ChainKey3["ARBITRUM"] = "arbitrum";
  ChainKey3["ARBITURM_GOERLI"] = "arbitrum-goerli";
  ChainKey3["OPTIMISM"] = "optimism";
  ChainKey3["OPTIMISM_GOERLI"] = "optimism-goerli";
  ChainKey3["FANTOM"] = "fantom";
  ChainKey3["FANTOM_TESTNET"] = "fantom-testnet";
  ChainKey3["SWIMMER"] = "swimmer";
  ChainKey3["SWIMMER_TESTNET"] = "swimmer-testnet";
  ChainKey3["DFK"] = "dfk";
  ChainKey3["DFK_TESTNET"] = "dfk-testnet";
  ChainKey3["HARMONY"] = "harmony";
  ChainKey3["HARMONY_TESTNET"] = "harmony-testnet";
  ChainKey3["DEXALOT"] = "dexalot";
  ChainKey3["DEVALOT_TESTNET"] = "dexalot-testnet";
  ChainKey3["CELO"] = "celo";
  ChainKey3["ALFAJORES"] = "alfajores";
  ChainKey3["MOONBEAM"] = "moonbeam";
  ChainKey3["MOONBASE"] = "moonbase";
  ChainKey3["MOONRIVER"] = "moonriver";
  ChainKey3["INTAIN"] = "intain";
  ChainKey3["INTAIN_TESTNET"] = "intain-testnet";
  ChainKey3["FUSE"] = "fuse";
  ChainKey3["FUSESPARK"] = "fusespark";
  ChainKey3["GNOSIS"] = "gnosis";
  ChainKey3["CHIADO"] = "chiado";
  ChainKey3["APTOS"] = "aptos";
  ChainKey3["APTOS_TESTNET"] = "aptos-testnet";
  ChainKey3["KLAYTN"] = "klaytn";
  ChainKey3["KLAYTN_BAOBAB"] = "klaytn-baobab";
  ChainKey3["METIS"] = "metis";
  ChainKey3["METIS_GOERLI"] = "metis-goerli";
  ChainKey3["COREDAO"] = "coredao";
  ChainKey3["COREDAO_TESTNET"] = "coredao-testnet";
  ChainKey3["METER"] = "meter";
  ChainKey3["METER_TESTNET"] = "meter-testnet";
  ChainKey3["OKX"] = "okx";
  ChainKey3["OKX_TESTNET"] = "okx-testnet";
  ChainKey3["ZKSYNC"] = "zksync";
  ChainKey3["ZKSYNC_TESTNET"] = "zksync-testnet";
  ChainKey3["SHRAPNEL"] = "shrapnel";
  ChainKey3["SHRAPNEL_TESTNET"] = "shrapnel-testnet";
  ChainKey3["DOS"] = "dos";
  ChainKey3["DOS_TESTNET"] = "dos-testnet";
  ChainKey3["ZKEVM"] = "zkevm";
  ChainKey3["ZKEVM_TESTNET"] = "zkevm-testnet";
  ChainKey3["NOVA"] = "nova";
  ChainKey3["TENET"] = "tenet";
  ChainKey3["TENET_TESTNET"] = "tenet-testnet";
  ChainKey3["CANTO"] = "canto";
  ChainKey3["CANTO_TESTNET"] = "canto-testnet";
  ChainKey3["KAVA"] = "kava";
  ChainKey3["KAVA_TESTNET"] = "kava-testnet";
  ChainKey3["SOLANA"] = "solana";
  ChainKey3["SOLANA_TESTNET"] = "solana-testnet";
  ChainKey3["LINEA_GOERLI"] = "linea-goerli";
  ChainKey3["BASE_GOERLI"] = "base-goerli";
  ChainKey3["AAVEGOTCHI_TESTNET"] = "aavegotchi-testnet";
  ChainKey3["SCROLL"] = "scroll";
  ChainKey3["SCROLL_TESTNET"] = "scroll-testnet";
  ChainKey3["CATHAY_TESTNET"] = "cathay-testnet";
  ChainKey3["ORDERLY"] = "orderly";
  ChainKey3["ORDERLY_TESTNET"] = "orderly-testnet";
  ChainKey3["BLOCKGEN_TESTNET"] = "blockgen-testnet";
  ChainKey3["BEAM_TESTNET"] = "beam-testnet";
  ChainKey3["BEAM"] = "beam";
  ChainKey3["MANTLE"] = "mantle";
  ChainKey3["MANTLE_TESTNET"] = "mantle-testnet";
  ChainKey3["HUBBLE"] = "hubble";
  ChainKey3["HUBBLE_TESTNET"] = "hubble-testnet";
  ChainKey3["LINEA"] = "linea";
  ChainKey3["BASE"] = "base";
  ChainKey3["ZORA_TESTNET"] = "zora-testnet";
  ChainKey3["ZORA"] = "zora";
  ChainKey3["TOMO_TESTNET"] = "tomo-testnet";
  ChainKey3["TOMO"] = "tomo";
  ChainKey3["LOOT_TESTNET"] = "loot-testnet";
  ChainKey3["LOOT"] = "loot";
  ChainKey3["TELOS_TESTNET"] = "telos-testnet";
  ChainKey3["TELOS"] = "telos";
  ChainKey3["AURORA"] = "aurora";
  ChainKey3["AURORA_TESTNET"] = "aurora-testnet";
  ChainKey3["OPBNB"] = "opbnb";
  ChainKey3["OPBNB_TESTNET"] = "opbnb-testnet";
  ChainKey3["SHIMMER"] = "shimmer";
  ChainKey3["SHIMMER_TESTNET"] = "shimmer-testnet";
  ChainKey3["MONAD_TESTNET"] = "monad-testnet";
  ChainKey3["LIF3"] = "lif3";
  ChainKey3["LIF3_TESTNET"] = "lif3-testnet";
  ChainKey3["SPRUCE_TESTNET"] = "spruce-testnet";
  ChainKey3["PGJJTK_TESTNET"] = "pgjjtk-testnet";
  ChainKey3["ODA_TESTNET"] = "oda-testnet";
  ChainKey3["KIWI_TESTNET"] = "kiwi-testnet";
  ChainKey3["ASTAR"] = "astar";
  ChainKey3["ASTAR_TESTNET"] = "astar-testnet";
  ChainKey3["CONFLUX"] = "conflux";
  ChainKey3["CONFLUX_TESTNET"] = "conflux-testnet";
  ChainKey3["EON"] = "eon";
  ChainKey3["EON_TESTNET"] = "eon-testnet";
  ChainKey3["XPLA"] = "xpla";
  ChainKey3["XPLA_TESTNET"] = "xpla-testnet";
  ChainKey3["HOLESKY"] = "holesky";
  ChainKey3["INJECTIVE_TESTNET"] = "injective-testnet";
  ChainKey3["IDEX_TESTNET"] = "idex-testnet";
  ChainKey3["ZKATANA_TESTNET"] = "zkatana-testnet";
  ChainKey3["MANTA"] = "manta";
  ChainKey3["MANTA_TESTNET"] = "manta-testnet";
  ChainKey3["FRAME_TESTNET"] = "frame-testnet";
  ChainKey3["PGN"] = "pgn";
  ChainKey3["PGN_TESTNET"] = "pgn-testnet";
  ChainKey3["POLYGONCDK_TESTNET"] = "polygoncdk-testnet";
  ChainKey3["ARBSEP_TESTNET"] = "arbsep-testnet";
  ChainKey3["OPTSEP_TESTNET"] = "optsep-testnet";
  return ChainKey3;
})(ChainKey || {});

// src/constants/chainKey.ts
var CHAIN_KEY = {
  [EndpointId.ETHEREUM_MAINNET]: "ethereum" /* ETHEREUM */,
  [EndpointId.BSC_MAINNET]: "bsc" /* BSC */,
  [EndpointId.AVALANCHE_MAINNET]: "avalanche" /* AVALANCHE */,
  [EndpointId.POLYGON_MAINNET]: "polygon" /* POLYGON */,
  [EndpointId.ARBITRUM_MAINNET]: "arbitrum" /* ARBITRUM */,
  [EndpointId.OPTIMISM_MAINNET]: "optimism" /* OPTIMISM */,
  [EndpointId.FANTOM_MAINNET]: "fantom" /* FANTOM */,
  [EndpointId.SWIMMER_MAINNET]: "swimmer" /* SWIMMER */,
  [EndpointId.DFK_MAINNET]: "dfk" /* DFK */,
  [EndpointId.HARMONY_MAINNET]: "harmony" /* HARMONY */,
  [EndpointId.DEXALOT_MAINNET]: "dexalot" /* DEXALOT */,
  [EndpointId.CELO_MAINNET]: "celo" /* CELO */,
  [EndpointId.MOONBEAM_MAINNET]: "moonbeam" /* MOONBEAM */,
  [EndpointId.INTAIN_MAINNET]: "intain" /* INTAIN */,
  [EndpointId.FUSE_MAINNET]: "fuse" /* FUSE */,
  [EndpointId.GNOSIS_MAINNET]: "gnosis" /* GNOSIS */,
  [EndpointId.APTOS_MAINNET]: "aptos" /* APTOS */,
  [EndpointId.KLAYTN_MAINNET]: "klaytn" /* KLAYTN */,
  [EndpointId.METIS_MAINNET]: "metis" /* METIS */,
  [EndpointId.COREDAO_MAINNET]: "coredao" /* COREDAO */,
  [EndpointId.METER_MAINNET]: "meter" /* METER */,
  [EndpointId.OKX_MAINNET]: "okx" /* OKX */,
  [EndpointId.GOERLI_MAINNET]: "goerli" /* GOERLI */,
  [EndpointId.ZKSYNC_MAINNET]: "zksync" /* ZKSYNC */,
  [EndpointId.SHRAPNEL_MAINNET]: "shrapnel" /* SHRAPNEL */,
  [EndpointId.DOS_MAINNET]: "dos" /* DOS */,
  [EndpointId.ZKPOLYGON_MAINNET]: "zkevm" /* ZKEVM */,
  [EndpointId.SEPOLIA_MAINNET]: "sepolia" /* SEPOLIA */,
  [EndpointId.MOONRIVER_MAINNET]: "moonriver" /* MOONRIVER */,
  [EndpointId.NOVA_MAINNET]: "nova" /* NOVA */,
  [EndpointId.TENET_MAINNET]: "tenet" /* TENET */,
  [EndpointId.CANTO_MAINNET]: "canto" /* CANTO */,
  [EndpointId.KAVA_MAINNET]: "kava" /* KAVA */,
  [EndpointId.MANTLE_MAINNET]: "mantle" /* MANTLE */,
  [EndpointId.HUBBLE_MAINNET]: "hubble" /* HUBBLE */,
  [EndpointId.ZKCONSENSYS_MAINNET]: "linea" /* LINEA */,
  [EndpointId.BASE_MAINNET]: "base" /* BASE */,
  [EndpointId.ZORA_MAINNET]: "zora" /* ZORA */,
  [EndpointId.TOMO_MAINNET]: "tomo" /* TOMO */,
  [EndpointId.LOOT_MAINNET]: "loot" /* LOOT */,
  [EndpointId.TELOS_MAINNET]: "telos" /* TELOS */,
  [EndpointId.MERITCIRCLE_MAINNET]: "beam" /* BEAM */,
  [EndpointId.OPBNB_MAINNET]: "opbnb" /* OPBNB */,
  [EndpointId.LIF3_MAINNET]: "lif3" /* LIF3 */,
  [EndpointId.ASTAR_MAINNET]: "astar" /* ASTAR */,
  [EndpointId.AURORA_MAINNET]: "aurora" /* AURORA */,
  [EndpointId.CONFLUX_MAINNET]: "conflux" /* CONFLUX */,
  [EndpointId.ORDERLY_MAINNET]: "orderly" /* ORDERLY */,
  [EndpointId.SCROLL_MAINNET]: "scroll" /* SCROLL */,
  [EndpointId.EON_MAINNET]: "eon" /* EON */,
  [EndpointId.XPLA_MAINNET]: "xpla" /* XPLA */,
  [EndpointId.MANTA_MAINNET]: "manta" /* MANTA */,
  [EndpointId.PGN_MAINNET]: "pgn" /* PGN */,
  [EndpointId.SHIMMER_MAINNET]: "shimmer" /* SHIMMER */,
  [EndpointId.SOLANA_MAINNET]: "solana" /* SOLANA */,
  // v2 mainnet
  [EndpointId.ETHEREUM_V2_MAINNET]: "ethereum" /* ETHEREUM */,
  [EndpointId.BSC_V2_MAINNET]: "bsc" /* BSC */,
  [EndpointId.AVALANCHE_V2_MAINNET]: "avalanche" /* AVALANCHE */,
  [EndpointId.POLYGON_V2_MAINNET]: "polygon" /* POLYGON */,
  [EndpointId.ARBITRUM_V2_MAINNET]: "arbitrum" /* ARBITRUM */,
  [EndpointId.OPTIMISM_V2_MAINNET]: "optimism" /* OPTIMISM */,
  [EndpointId.FANTOM_V2_MAINNET]: "fantom" /* FANTOM */,
  [EndpointId.SWIMMER_V2_MAINNET]: "swimmer" /* SWIMMER */,
  [EndpointId.DFK_V2_MAINNET]: "dfk" /* DFK */,
  [EndpointId.HARMONY_V2_MAINNET]: "harmony" /* HARMONY */,
  [EndpointId.DEXALOT_V2_MAINNET]: "dexalot" /* DEXALOT */,
  [EndpointId.CELO_V2_MAINNET]: "celo" /* CELO */,
  [EndpointId.MOONBEAM_V2_MAINNET]: "moonbeam" /* MOONBEAM */,
  [EndpointId.INTAIN_V2_MAINNET]: "intain" /* INTAIN */,
  [EndpointId.FUSE_V2_MAINNET]: "fuse" /* FUSE */,
  [EndpointId.GNOSIS_V2_MAINNET]: "gnosis" /* GNOSIS */,
  [EndpointId.APTOS_V2_MAINNET]: "aptos" /* APTOS */,
  [EndpointId.KLAYTN_V2_MAINNET]: "klaytn" /* KLAYTN */,
  [EndpointId.METIS_V2_MAINNET]: "metis" /* METIS */,
  [EndpointId.COREDAO_V2_MAINNET]: "coredao" /* COREDAO */,
  [EndpointId.GOERLI_V2_MAINNET]: "goerli" /* GOERLI */,
  [EndpointId.OKX_V2_MAINNET]: "okx" /* OKX */,
  [EndpointId.CANTO_V2_MAINNET]: "canto" /* CANTO */,
  [EndpointId.ZKSYNC_V2_MAINNET]: "zksync" /* ZKSYNC */,
  [EndpointId.SHRAPNEL_V2_MAINNET]: "shrapnel" /* SHRAPNEL */,
  [EndpointId.DOS_V2_MAINNET]: "dos" /* DOS */,
  [EndpointId.ZKPOLYGON_V2_MAINNET]: "zkevm" /* ZKEVM */,
  [EndpointId.SEPOLIA_V2_MAINNET]: "sepolia" /* SEPOLIA */,
  [EndpointId.MOONRIVER_V2_MAINNET]: "moonriver" /* MOONRIVER */,
  [EndpointId.SOLANA_V2_MAINNET]: "solana" /* SOLANA */,
  [EndpointId.TENET_V2_MAINNET]: "tenet" /* TENET */,
  [EndpointId.NOVA_V2_MAINNET]: "nova" /* NOVA */,
  [EndpointId.METER_V2_MAINNET]: "meter" /* METER */,
  [EndpointId.KAVA_V2_MAINNET]: "kava" /* KAVA */,
  [EndpointId.MANTLE_V2_MAINNET]: "mantle" /* MANTLE */,
  [EndpointId.HUBBLE_V2_MAINNET]: "hubble" /* HUBBLE */,
  [EndpointId.ZKCONSENSYS_V2_MAINNET]: "linea" /* LINEA */,
  [EndpointId.BASE_V2_MAINNET]: "base" /* BASE */,
  [EndpointId.ZORA_V2_MAINNET]: "zora" /* ZORA */,
  [EndpointId.TOMO_V2_MAINNET]: "tomo" /* TOMO */,
  [EndpointId.LOOT_V2_MAINNET]: "loot" /* LOOT */,
  [EndpointId.TELOS_V2_MAINNET]: "telos" /* TELOS */,
  [EndpointId.MERITCIRCLE_V2_MAINNET]: "beam" /* BEAM */,
  [EndpointId.OPBNB_V2_MAINNET]: "opbnb" /* OPBNB */,
  [EndpointId.LIF3_V2_MAINNET]: "lif3" /* LIF3 */,
  [EndpointId.ASTAR_V2_MAINNET]: "astar" /* ASTAR */,
  [EndpointId.AURORA_V2_MAINNET]: "aurora" /* AURORA */,
  [EndpointId.CONFLUX_V2_MAINNET]: "conflux" /* CONFLUX */,
  [EndpointId.ORDERLY_V2_MAINNET]: "orderly" /* ORDERLY */,
  [EndpointId.SCROLL_V2_MAINNET]: "scroll" /* SCROLL */,
  [EndpointId.EON_V2_MAINNET]: "eon" /* EON */,
  [EndpointId.XPLA_V2_MAINNET]: "xpla" /* XPLA */,
  [EndpointId.MANTA_V2_MAINNET]: "manta" /* MANTA */,
  [EndpointId.PGN_V2_MAINNET]: "pgn" /* PGN */,
  [EndpointId.SHIMMER_V2_MAINNET]: "shimmer" /* SHIMMER */,
  // testnet
  [EndpointId.ETHEREUM_TESTNET]: "goerli" /* GOERLI */,
  [EndpointId.BSC_TESTNET]: "bsc-testnet" /* BSC_TESTNET */,
  [EndpointId.POLYGON_TESTNET]: "mumbai" /* MUMBAI */,
  [EndpointId.FANTOM_TESTNET]: "fantom-testnet" /* FANTOM_TESTNET */,
  [EndpointId.HARMONY_TESTNET]: "harmony-testnet" /* HARMONY_TESTNET */,
  [EndpointId.AVALANCHE_TESTNET]: "fuji" /* FUJI */,
  [EndpointId.ARBITRUM_TESTNET]: "arbitrum-goerli" /* ARBITURM_GOERLI */,
  [EndpointId.OPTIMISM_TESTNET]: "optimism-goerli" /* OPTIMISM_GOERLI */,
  [EndpointId.DFK_TESTNET]: "dfk-testnet" /* DFK_TESTNET */,
  [EndpointId.SWIMMER_TESTNET]: "swimmer-testnet" /* SWIMMER_TESTNET */,
  [EndpointId.DEXALOT_TESTNET]: "dexalot-testnet" /* DEVALOT_TESTNET */,
  [EndpointId.GNOSIS_TESTNET]: "chiado" /* CHIADO */,
  [EndpointId.CELO_TESTNET]: "alfajores" /* ALFAJORES */,
  [EndpointId.APTOS_TESTNET]: "aptos-testnet" /* APTOS_TESTNET */,
  [EndpointId.MOONBEAM_TESTNET]: "moonbase" /* MOONBASE */,
  [EndpointId.FUSE_TESTNET]: "fusespark" /* FUSESPARK */,
  [EndpointId.ZKSYNC_TESTNET]: "zksync-testnet" /* ZKSYNC_TESTNET */,
  [EndpointId.SHRAPNEL_TESTNET]: "shrapnel-testnet" /* SHRAPNEL_TESTNET */,
  [EndpointId.DOS_TESTNET]: "dos-testnet" /* DOS_TESTNET */,
  [EndpointId.KLAYTN_TESTNET]: "klaytn-baobab" /* KLAYTN_BAOBAB */,
  [EndpointId.METIS_TESTNET]: "metis-goerli" /* METIS_GOERLI */,
  [EndpointId.INTAIN_TESTNET]: "intain-testnet" /* INTAIN_TESTNET */,
  [EndpointId.COREDAO_TESTNET]: "coredao-testnet" /* COREDAO_TESTNET */,
  [EndpointId.ZKPOLYGON_TESTNET]: "zkevm-testnet" /* ZKEVM_TESTNET */,
  [EndpointId.OKX_TESTNET]: "okx-testnet" /* OKX_TESTNET */,
  [EndpointId.SOLANA_TESTNET]: "solana-testnet" /* SOLANA_TESTNET */,
  [EndpointId.METER_TESTNET]: "meter-testnet" /* METER_TESTNET */,
  [EndpointId.ZKCONSENSYS_TESTNET]: "linea-goerli" /* LINEA_GOERLI */,
  [EndpointId.SEPOLIA_TESTNET]: "sepolia" /* SEPOLIA */,
  [EndpointId.BASE_TESTNET]: "base-goerli" /* BASE_GOERLI */,
  [EndpointId.AAVEGOTCHI_TESTNET]: "aavegotchi-testnet" /* AAVEGOTCHI_TESTNET */,
  [EndpointId.SCROLL_TESTNET]: "scroll-testnet" /* SCROLL_TESTNET */,
  [EndpointId.CATHAY_TESTNET]: "cathay-testnet" /* CATHAY_TESTNET */,
  [EndpointId.KAVA_TESTNET]: "kava-testnet" /* KAVA_TESTNET */,
  [EndpointId.TENET_TESTNET]: "tenet-testnet" /* TENET_TESTNET */,
  [EndpointId.CANTO_TESTNET]: "canto-testnet" /* CANTO_TESTNET */,
  [EndpointId.ORDERLY_TESTNET]: "orderly-testnet" /* ORDERLY_TESTNET */,
  [EndpointId.BLOCKGEN_TESTNET]: "blockgen-testnet" /* BLOCKGEN_TESTNET */,
  [EndpointId.MERITCIRCLE_TESTNET]: "beam-testnet" /* BEAM_TESTNET */,
  [EndpointId.MANTLE_TESTNET]: "mantle-testnet" /* MANTLE_TESTNET */,
  [EndpointId.HUBBLE_TESTNET]: "hubble-testnet" /* HUBBLE_TESTNET */,
  [EndpointId.ZORA_TESTNET]: "zora-testnet" /* ZORA_TESTNET */,
  [EndpointId.TOMO_TESTNET]: "tomo-testnet" /* TOMO_TESTNET */,
  [EndpointId.LOOT_TESTNET]: "loot-testnet" /* LOOT_TESTNET */,
  [EndpointId.TELOS_TESTNET]: "telos-testnet" /* TELOS_TESTNET */,
  [EndpointId.AURORA_TESTNET]: "aurora-testnet" /* AURORA_TESTNET */,
  [EndpointId.OPBNB_TESTNET]: "opbnb-testnet" /* OPBNB_TESTNET */,
  [EndpointId.SHIMMER_TESTNET]: "shimmer-testnet" /* SHIMMER_TESTNET */,
  [EndpointId.MONAD_TESTNET]: "monad-testnet" /* MONAD_TESTNET */,
  [EndpointId.LIF3_TESTNET]: "lif3-testnet" /* LIF3_TESTNET */,
  [EndpointId.SPRUCE_TESTNET]: "spruce-testnet" /* SPRUCE_TESTNET */,
  [EndpointId.PGJJTK_TESTNET]: "pgjjtk-testnet" /* PGJJTK_TESTNET */,
  [EndpointId.ODA_TESTNET]: "oda-testnet" /* ODA_TESTNET */,
  [EndpointId.KIWI_TESTNET]: "kiwi-testnet" /* KIWI_TESTNET */,
  [EndpointId.ASTAR_TESTNET]: "astar-testnet" /* ASTAR_TESTNET */,
  [EndpointId.CONFLUX_TESTNET]: "conflux-testnet" /* CONFLUX_TESTNET */,
  [EndpointId.EON_TESTNET]: "eon-testnet" /* EON_TESTNET */,
  [EndpointId.XPLA_TESTNET]: "xpla-testnet" /* XPLA_TESTNET */,
  [EndpointId.HOLESKY_TESTNET]: "holesky" /* HOLESKY */,
  [EndpointId.INJECTIVE_TESTNET]: "injective-testnet" /* INJECTIVE_TESTNET */,
  [EndpointId.IDEX_TESTNET]: "idex-testnet" /* IDEX_TESTNET */,
  [EndpointId.ZKATANA_TESTNET]: "zkatana-testnet" /* ZKATANA_TESTNET */,
  [EndpointId.MANTA_TESTNET]: "manta-testnet" /* MANTA_TESTNET */,
  [EndpointId.FRAME_TESTNET]: "frame-testnet" /* FRAME_TESTNET */,
  [EndpointId.PGN_TESTNET]: "pgn-testnet" /* PGN_TESTNET */,
  [EndpointId.POLYGONCDK_TESTNET]: "polygoncdk-testnet" /* POLYGONCDK_TESTNET */,
  [EndpointId.ARBSEP_TESTNET]: "arbsep-testnet" /* ARBSEP_TESTNET */,
  [EndpointId.OPTSEP_TESTNET]: "optsep-testnet" /* OPTSEP_TESTNET */,
  // v2 testnet
  [EndpointId.ETHEREUM_V2_TESTNET]: "goerli" /* GOERLI */,
  [EndpointId.POLYGON_V2_TESTNET]: "mumbai" /* MUMBAI */,
  [EndpointId.SOLANA_V2_TESTNET]: "solana-testnet" /* SOLANA_TESTNET */,
  [EndpointId.ARBITRUM_V2_TESTNET]: "arbitrum-goerli" /* ARBITURM_GOERLI */,
  [EndpointId.OPTIMISM_V2_TESTNET]: "optimism-goerli" /* OPTIMISM_GOERLI */,
  [EndpointId.AVALANCHE_V2_TESTNET]: "fuji" /* FUJI */,
  [EndpointId.BSC_V2_TESTNET]: "bsc-testnet" /* BSC_TESTNET */,
  [EndpointId.FANTOM_V2_TESTNET]: "fantom-testnet" /* FANTOM_TESTNET */,
  [EndpointId.DFK_V2_TESTNET]: "dfk-testnet" /* DFK_TESTNET */,
  [EndpointId.HARMONY_V2_TESTNET]: "harmony-testnet" /* HARMONY_TESTNET */,
  [EndpointId.DEXALOT_V2_TESTNET]: "dexalot-testnet" /* DEVALOT_TESTNET */,
  [EndpointId.CELO_V2_TESTNET]: "alfajores" /* ALFAJORES */,
  [EndpointId.MOONBEAM_V2_TESTNET]: "moonbase" /* MOONBASE */,
  [EndpointId.FUSE_V2_TESTNET]: "fusespark" /* FUSESPARK */,
  [EndpointId.GNOSIS_V2_TESTNET]: "chiado" /* CHIADO */,
  [EndpointId.ZKSYNC_V2_TESTNET]: "zksync-testnet" /* ZKSYNC_TESTNET */,
  [EndpointId.SHRAPNEL_V2_TESTNET]: "shrapnel-testnet" /* SHRAPNEL_TESTNET */,
  [EndpointId.DOS_V2_TESTNET]: "dos-testnet" /* DOS_TESTNET */,
  [EndpointId.KLAYTN_V2_TESTNET]: "klaytn-baobab" /* KLAYTN_BAOBAB */,
  [EndpointId.METIS_V2_TESTNET]: "metis-goerli" /* METIS_GOERLI */,
  [EndpointId.COREDAO_V2_TESTNET]: "coredao-testnet" /* COREDAO_TESTNET */,
  [EndpointId.OKX_V2_TESTNET]: "okx-testnet" /* OKX_TESTNET */,
  [EndpointId.METER_V2_TESTNET]: "meter-testnet" /* METER_TESTNET */,
  [EndpointId.ZKCONSENSYS_V2_TESTNET]: "linea-goerli" /* LINEA_GOERLI */,
  [EndpointId.ZKPOLYGON_V2_TESTNET]: "zkevm-testnet" /* ZKEVM_TESTNET */,
  [EndpointId.CANTO_V2_TESTNET]: "canto-testnet" /* CANTO_TESTNET */,
  [EndpointId.BASE_V2_TESTNET]: "base-goerli" /* BASE_GOERLI */,
  [EndpointId.SEPOLIA_V2_TESTNET]: "sepolia" /* SEPOLIA */,
  [EndpointId.AAVEGOTCHI_V2_TESTNET]: "aavegotchi-testnet" /* AAVEGOTCHI_TESTNET */,
  [EndpointId.SCROLL_V2_TESTNET]: "scroll-testnet" /* SCROLL_TESTNET */,
  [EndpointId.CATHAY_V2_TESTNET]: "cathay-testnet" /* CATHAY_TESTNET */,
  [EndpointId.KAVA_V2_TESTNET]: "kava-testnet" /* KAVA_TESTNET */,
  [EndpointId.TENET_V2_TESTNET]: "tenet-testnet" /* TENET_TESTNET */,
  [EndpointId.ORDERLY_V2_TESTNET]: "orderly-testnet" /* ORDERLY_TESTNET */,
  [EndpointId.BLOCKGEN_V2_TESTNET]: "blockgen-testnet" /* BLOCKGEN_TESTNET */,
  [EndpointId.MERITCIRCLE_V2_TESTNET]: "beam-testnet" /* BEAM_TESTNET */,
  [EndpointId.MANTLE_V2_TESTNET]: "mantle-testnet" /* MANTLE_TESTNET */,
  [EndpointId.HUBBLE_V2_TESTNET]: "hubble-testnet" /* HUBBLE_TESTNET */,
  [EndpointId.ZORA_V2_TESTNET]: "zora-testnet" /* ZORA_TESTNET */,
  [EndpointId.TOMO_V2_TESTNET]: "tomo-testnet" /* TOMO_TESTNET */,
  [EndpointId.LOOT_V2_TESTNET]: "loot-testnet" /* LOOT_TESTNET */,
  [EndpointId.TELOS_V2_TESTNET]: "telos-testnet" /* TELOS_TESTNET */,
  [EndpointId.AURORA_V2_TESTNET]: "aurora-testnet" /* AURORA_TESTNET */,
  [EndpointId.OPBNB_V2_TESTNET]: "opbnb-testnet" /* OPBNB_TESTNET */,
  [EndpointId.SHIMMER_V2_TESTNET]: "shimmer-testnet" /* SHIMMER_TESTNET */,
  [EndpointId.MONAD_V2_TESTNET]: "monad-testnet" /* MONAD_TESTNET */,
  [EndpointId.LIF3_V2_TESTNET]: "lif3-testnet" /* LIF3_TESTNET */,
  [EndpointId.SPRUCE_V2_TESTNET]: "spruce-testnet" /* SPRUCE_TESTNET */,
  [EndpointId.PGJJTK_V2_TESTNET]: "pgjjtk-testnet" /* PGJJTK_TESTNET */,
  [EndpointId.ODA_V2_TESTNET]: "oda-testnet" /* ODA_TESTNET */,
  [EndpointId.KIWI_V2_TESTNET]: "kiwi-testnet" /* KIWI_TESTNET */,
  [EndpointId.ASTAR_V2_TESTNET]: "astar-testnet" /* ASTAR_TESTNET */,
  [EndpointId.CONFLUX_V2_TESTNET]: "conflux-testnet" /* CONFLUX_TESTNET */,
  [EndpointId.EON_V2_TESTNET]: "eon-testnet" /* EON_TESTNET */,
  [EndpointId.XPLA_V2_TESTNET]: "xpla-testnet" /* XPLA_TESTNET */,
  [EndpointId.HOLESKY_V2_TESTNET]: "holesky" /* HOLESKY */,
  [EndpointId.INJECTIVE_V2_TESTNET]: "injective-testnet" /* INJECTIVE_TESTNET */,
  [EndpointId.IDEX_V2_TESTNET]: "idex-testnet" /* IDEX_TESTNET */,
  [EndpointId.ZKATANA_V2_TESTNET]: "zkatana-testnet" /* ZKATANA_TESTNET */,
  [EndpointId.MANTA_V2_TESTNET]: "manta-testnet" /* MANTA_TESTNET */,
  [EndpointId.FRAME_V2_TESTNET]: "frame-testnet" /* FRAME_TESTNET */,
  [EndpointId.PGN_V2_TESTNET]: "pgn-testnet" /* PGN_TESTNET */,
  [EndpointId.POLYGONCDK_V2_TESTNET]: "polygoncdk-testnet" /* POLYGONCDK_TESTNET */,
  [EndpointId.ARBSEP_V2_TESTNET]: "arbsep-testnet" /* ARBSEP_TESTNET */,
  [EndpointId.OPTSEP_V2_TESTNET]: "optsep-testnet" /* OPTSEP_TESTNET */,
  // sandbox
  [EndpointId.ETHEREUM_SANDBOX]: "goerli" /* GOERLI */,
  [EndpointId.BSC_SANDBOX]: "bsc-testnet" /* BSC_TESTNET */,
  [EndpointId.POLYGON_SANDBOX]: "mumbai" /* MUMBAI */,
  [EndpointId.AVALANCHE_SANDBOX]: "fuji" /* FUJI */,
  [EndpointId.APTOS_SANDBOX]: "aptos" /* APTOS */,
  // v2 sandbox
  [EndpointId.ETHEREUM_V2_SANDBOX]: "goerli" /* GOERLI */,
  [EndpointId.POLYGON_V2_SANDBOX]: "mumbai" /* MUMBAI */,
  [EndpointId.SOLANA_V2_SANDBOX]: "solana-testnet" /* SOLANA_TESTNET */,
  [EndpointId.BSC_V2_SANDBOX]: "bsc-testnet" /* BSC_TESTNET */,
  [EndpointId.AVALANCHE_V2_SANDBOX]: "fuji" /* FUJI */
};

// src/constants/environment.ts
var ENVIRONMENT = {
  [EndpointId.ETHEREUM_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.BSC_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.AVALANCHE_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.POLYGON_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.ARBITRUM_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.OPTIMISM_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.FANTOM_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.SWIMMER_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.DFK_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.HARMONY_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.DEXALOT_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.CELO_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.MOONBEAM_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.INTAIN_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.FUSE_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.GNOSIS_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.APTOS_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.KLAYTN_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.METIS_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.COREDAO_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.METER_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.OKX_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.GOERLI_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.ZKSYNC_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.SHRAPNEL_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.DOS_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.ZKPOLYGON_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.SEPOLIA_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.MOONRIVER_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.NOVA_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.TENET_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.CANTO_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.KAVA_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.MANTLE_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.HUBBLE_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.ZKCONSENSYS_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.BASE_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.ZORA_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.TOMO_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.LOOT_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.TELOS_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.MERITCIRCLE_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.OPBNB_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.LIF3_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.ASTAR_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.AURORA_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.CONFLUX_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.ORDERLY_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.SCROLL_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.EON_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.XPLA_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.MANTA_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.PGN_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.SHIMMER_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.SOLANA_MAINNET]: "mainnet" /* MAINNET */,
  // v2 mainnet
  [EndpointId.ETHEREUM_V2_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.BSC_V2_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.AVALANCHE_V2_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.POLYGON_V2_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.ARBITRUM_V2_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.OPTIMISM_V2_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.FANTOM_V2_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.SWIMMER_V2_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.DFK_V2_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.HARMONY_V2_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.DEXALOT_V2_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.CELO_V2_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.MOONBEAM_V2_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.INTAIN_V2_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.FUSE_V2_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.GNOSIS_V2_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.APTOS_V2_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.KLAYTN_V2_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.METIS_V2_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.COREDAO_V2_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.GOERLI_V2_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.OKX_V2_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.CANTO_V2_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.ZKSYNC_V2_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.SHRAPNEL_V2_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.DOS_V2_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.ZKPOLYGON_V2_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.SEPOLIA_V2_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.MOONRIVER_V2_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.SOLANA_V2_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.TENET_V2_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.NOVA_V2_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.METER_V2_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.KAVA_V2_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.MANTLE_V2_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.HUBBLE_V2_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.ZKCONSENSYS_V2_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.BASE_V2_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.ZORA_V2_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.TOMO_V2_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.LOOT_V2_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.TELOS_V2_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.MERITCIRCLE_V2_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.OPBNB_V2_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.LIF3_V2_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.ASTAR_V2_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.AURORA_V2_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.CONFLUX_V2_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.ORDERLY_V2_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.SCROLL_V2_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.EON_V2_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.XPLA_V2_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.MANTA_V2_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.PGN_V2_MAINNET]: "mainnet" /* MAINNET */,
  [EndpointId.SHIMMER_V2_MAINNET]: "mainnet" /* MAINNET */,
  // testnet
  [EndpointId.ETHEREUM_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.BSC_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.POLYGON_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.FANTOM_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.HARMONY_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.AVALANCHE_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.ARBITRUM_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.OPTIMISM_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.DFK_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.SWIMMER_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.DEXALOT_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.GNOSIS_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.CELO_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.APTOS_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.MOONBEAM_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.FUSE_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.ZKSYNC_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.SHRAPNEL_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.DOS_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.KLAYTN_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.METIS_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.INTAIN_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.COREDAO_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.ZKPOLYGON_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.OKX_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.SOLANA_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.METER_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.ZKCONSENSYS_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.SEPOLIA_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.BASE_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.AAVEGOTCHI_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.SCROLL_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.CATHAY_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.KAVA_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.TENET_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.CANTO_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.ORDERLY_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.BLOCKGEN_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.MERITCIRCLE_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.MANTLE_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.HUBBLE_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.ZORA_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.TOMO_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.LOOT_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.TELOS_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.AURORA_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.OPBNB_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.SHIMMER_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.MONAD_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.LIF3_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.SPRUCE_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.PGJJTK_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.ODA_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.KIWI_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.ASTAR_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.CONFLUX_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.EON_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.XPLA_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.HOLESKY_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.INJECTIVE_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.IDEX_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.ZKATANA_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.MANTA_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.FRAME_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.PGN_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.POLYGONCDK_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.ARBSEP_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.OPTSEP_TESTNET]: "testnet" /* TESTNET */,
  // v2 testnet
  [EndpointId.ETHEREUM_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.POLYGON_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.SOLANA_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.ARBITRUM_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.OPTIMISM_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.AVALANCHE_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.BSC_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.FANTOM_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.DFK_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.HARMONY_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.DEXALOT_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.CELO_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.MOONBEAM_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.FUSE_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.GNOSIS_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.ZKSYNC_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.SHRAPNEL_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.DOS_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.KLAYTN_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.METIS_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.COREDAO_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.OKX_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.METER_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.ZKCONSENSYS_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.ZKPOLYGON_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.CANTO_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.BASE_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.SEPOLIA_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.AAVEGOTCHI_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.SCROLL_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.CATHAY_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.KAVA_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.TENET_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.ORDERLY_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.BLOCKGEN_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.MERITCIRCLE_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.MANTLE_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.HUBBLE_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.ZORA_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.TOMO_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.LOOT_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.TELOS_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.AURORA_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.OPBNB_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.SHIMMER_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.MONAD_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.LIF3_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.SPRUCE_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.PGJJTK_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.ODA_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.KIWI_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.ASTAR_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.CONFLUX_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.EON_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.XPLA_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.HOLESKY_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.INJECTIVE_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.IDEX_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.ZKATANA_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.MANTA_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.FRAME_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.PGN_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.POLYGONCDK_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.ARBSEP_V2_TESTNET]: "testnet" /* TESTNET */,
  [EndpointId.OPTSEP_V2_TESTNET]: "testnet" /* TESTNET */,
  // sandbox
  [EndpointId.ETHEREUM_SANDBOX]: "testnet" /* TESTNET */,
  [EndpointId.BSC_SANDBOX]: "testnet" /* TESTNET */,
  [EndpointId.POLYGON_SANDBOX]: "testnet" /* TESTNET */,
  [EndpointId.AVALANCHE_SANDBOX]: "testnet" /* TESTNET */,
  [EndpointId.APTOS_SANDBOX]: "devnet" /* DEVNET */,
  // v2 sandbox
  [EndpointId.ETHEREUM_V2_SANDBOX]: "testnet" /* TESTNET */,
  [EndpointId.POLYGON_V2_SANDBOX]: "devnet" /* DEVNET */,
  [EndpointId.SOLANA_V2_SANDBOX]: "devnet" /* DEVNET */,
  [EndpointId.BSC_V2_SANDBOX]: "devnet" /* DEVNET */,
  [EndpointId.AVALANCHE_V2_SANDBOX]: "devnet" /* DEVNET */
};

// src/constants/ulnv1.ts
var ULN_V1_CHAINS = [
  EndpointId.ETHEREUM_MAINNET,
  EndpointId.ETHEREUM_TESTNET,
  EndpointId.ETHEREUM_SANDBOX,
  EndpointId.BSC_MAINNET,
  EndpointId.BSC_TESTNET,
  EndpointId.BSC_SANDBOX,
  EndpointId.AVALANCHE_MAINNET,
  EndpointId.AVALANCHE_TESTNET,
  EndpointId.AVALANCHE_SANDBOX,
  EndpointId.POLYGON_MAINNET,
  EndpointId.POLYGON_TESTNET,
  EndpointId.POLYGON_SANDBOX,
  EndpointId.ARBITRUM_MAINNET,
  EndpointId.OPTIMISM_MAINNET,
  EndpointId.FANTOM_MAINNET,
  EndpointId.FANTOM_TESTNET,
  EndpointId.SWIMMER_MAINNET,
  EndpointId.SWIMMER_TESTNET,
  EndpointId.DFK_MAINNET,
  EndpointId.DFK_TESTNET,
  EndpointId.HARMONY_MAINNET,
  EndpointId.HARMONY_TESTNET,
  EndpointId.MOONBEAM_MAINNET,
  EndpointId.MOONBEAM_TESTNET,
  EndpointId.DEXALOT_TESTNET,
  EndpointId.CELO_TESTNET,
  EndpointId.INTAIN_TESTNET
];
function networkToEndpointId(network, version) {
  const name = network.replace("-local", "");
  const [chain, stage] = name.split("-");
  const key = version === "v1" /* V1 */ ? `${chain.toUpperCase()}_${stage.toUpperCase()}` : `${chain.toUpperCase()}_V2_${stage.toUpperCase()}`;
  invariant__default.default(key in EndpointId, `Invalid network name: ${network}, key: ${key}`);
  return EndpointId[key];
}
function networkToEnv(network, version) {
  if (network.includes("local") || network === "hardhat") {
    return "local" /* LOCAL */;
  }
  const endpointId = networkToEndpointId(network, version);
  return ENVIRONMENT[endpointId];
}
function networkToStage(network) {
  const name = network.replace("-local", "");
  const stage = name.split("-")[1];
  return Stage[stage.toUpperCase()];
}
function endpointIdToNetwork(endpointId, envOrIsLocal) {
  const key = EndpointId[endpointId];
  invariant__default.default(key, `Invalid endpointId: ${endpointId}`);
  const network = key.toLowerCase().replace(/_/g, "-").replace("-v2", "");
  if (envOrIsLocal === "local" /* LOCAL */ || envOrIsLocal === true) {
    return `${network}-local`;
  }
  return network;
}
function endpointIdToVersion(endpointId) {
  const key = EndpointId[endpointId];
  invariant__default.default(key, `Invalid endpointId: ${endpointId}`);
  return key.includes("V2") ? "v2" /* V2 */ : "v1" /* V1 */;
}
function endpointIdToChainKey(endpointId) {
  const chainKey = CHAIN_KEY[endpointId];
  invariant__default.default(chainKey, `No ChainKey for EndpointId: ${endpointId}`);
  return chainKey;
}
function chainAndStageToEndpointId(chain, stage, version) {
  const key = version === "v2" /* V2 */ ? `${chain.toUpperCase()}_V2_${stage.toUpperCase()}` : `${chain.toUpperCase()}_${stage.toUpperCase()}`;
  invariant__default.default(key in EndpointId, `Invalid key: ${key}`);
  return EndpointId[key];
}
function chainAndStageToNetwork(chain, stage, envOrIsLocal) {
  if (envOrIsLocal === "local" /* LOCAL */ || envOrIsLocal === true) {
    return `${chain}-${stage}-local`;
  }
  return `${chain}-${stage}`;
}
function endpointDomainToNetwork(domain) {
  return chainAndStageToNetwork(domain.chain, domain.stage, domain.isLocal ? "local" /* LOCAL */ : void 0);
}
function endpointDomainToEndpointId(domain) {
  return chainAndStageToEndpointId(domain.chain, domain.stage, domain.version);
}
function endpointIdToEndpointDomain(endpointId, isLocal = false) {
  const chain = endpointIdToChain(endpointId);
  const stage = endpointIdToStage(endpointId);
  const version = endpointIdToVersion(endpointId);
  return { chain, stage, version, isLocal };
}
function networkToEndpointDomain(network, version) {
  const chain = networkToChain(network);
  const stage = networkToStage(network);
  const isLocal = network.endsWith("-local");
  return { chain, stage, version, isLocal };
}
function networkToChain(network) {
  return network.split("-")[0];
}
function networkToChainType(network) {
  const chain = networkToChain(network);
  return getChainType(chain);
}
function getChainType(chain) {
  switch (chain) {
    case "aptos" /* APTOS */:
      return "aptos" /* APTOS */;
    case "solana" /* SOLANA */:
      return "solana" /* SOLANA */;
    default:
      if (Object.values(Chain).includes(chain)) {
        return "evm" /* EVM */;
      } else {
        return "unknown" /* UNKNOWN */;
      }
  }
}
function endpointIdToChain(endpointId) {
  const key = EndpointId[endpointId];
  invariant__default.default(key, `Invalid endpointId: ${endpointId}`);
  return key.split("_")[0].toLowerCase();
}
function endpointIdToStage(endpointId) {
  let key = EndpointId[endpointId];
  invariant__default.default(key, `Invalid endpointId: ${endpointId}`);
  key = key.replace("_V2", "");
  return key.split("_")[1].toLowerCase();
}
function endpointIdToChainType(endpointId) {
  const chain = endpointIdToChain(endpointId);
  return getChainType(chain);
}
function getNetworksForStage(stage) {
  const networks = [];
  for (const key in EndpointId) {
    if (Number(key) >= 0) {
      const network = endpointIdToNetwork(Number(key));
      const s = network.split("-")[1];
      if (stage === s) {
        networks.push(network);
      }
    }
  }
  return networks;
}
var ULN_V1_BIAS = 100;
var getEndpointVersionForUlnVersion = (ulnVersion) => {
  const [
    ,
    ,
    /*majorVerion*/
    /*minorVersion*/
    endpointVersion
  ] = ulnVersion.split(".");
  if (!endpointVersion) {
    return "v1" /* V1 */;
  }
  const version = {
    "1": "v1" /* V1 */,
    "2": "v2" /* V2 */
  }[endpointVersion];
  if (!version) {
    throw new Error(`Invalid ulnVersion: ${ulnVersion}`);
  }
  return version;
};
function getChainIdForNetwork(chain, stage, ulnVersion) {
  const endpointId = chainAndStageToEndpointId(
    chain,
    stage,
    getEndpointVersionForUlnVersion(ulnVersion)
  );
  return (ulnVersion == "1" ? endpointId - ULN_V1_BIAS : endpointId).toString();
}
function getNetworkForChainId(targetChainId) {
  if (ULN_V1_CHAINS.includes(targetChainId + ULN_V1_BIAS)) {
    const endpointId = targetChainId + ULN_V1_BIAS;
    const chain2 = endpointIdToChain(endpointId);
    const stage2 = networkToStage(endpointIdToNetwork(endpointId));
    return {
      chainName: chain2,
      env: stage2,
      ulnVersion: "1"
    };
  }
  const chain = endpointIdToChain(targetChainId);
  const stage = networkToStage(endpointIdToNetwork(targetChainId));
  return {
    chainName: chain,
    env: stage,
    ulnVersion: "2"
  };
}

exports.CHAIN_KEY = CHAIN_KEY;
exports.Chain = Chain;
exports.ChainKey = ChainKey;
exports.ChainType = ChainType;
exports.ENVIRONMENT = ENVIRONMENT;
exports.EndpointId = EndpointId;
exports.EndpointVersion = EndpointVersion;
exports.Environment = Environment;
exports.MainnetEndpointId = MainnetEndpointId;
exports.MainnetV2EndpointId = MainnetV2EndpointId;
exports.MessageLibVersionV1 = MessageLibVersionV1;
exports.MessageLibVersionV2 = MessageLibVersionV2;
exports.SandboxEndpointId = SandboxEndpointId;
exports.SandboxV2EndpointId = SandboxV2EndpointId;
exports.Stage = Stage;
exports.TestnetEndpointId = TestnetEndpointId;
exports.TestnetV2EndpointId = TestnetV2EndpointId;
exports.ULN_V1_CHAINS = ULN_V1_CHAINS;
exports.chainAndStageToEndpointId = chainAndStageToEndpointId;
exports.chainAndStageToNetwork = chainAndStageToNetwork;
exports.endpointDomainToEndpointId = endpointDomainToEndpointId;
exports.endpointDomainToNetwork = endpointDomainToNetwork;
exports.endpointIdToChain = endpointIdToChain;
exports.endpointIdToChainKey = endpointIdToChainKey;
exports.endpointIdToChainType = endpointIdToChainType;
exports.endpointIdToEndpointDomain = endpointIdToEndpointDomain;
exports.endpointIdToNetwork = endpointIdToNetwork;
exports.endpointIdToStage = endpointIdToStage;
exports.endpointIdToVersion = endpointIdToVersion;
exports.getChainIdForNetwork = getChainIdForNetwork;
exports.getChainType = getChainType;
exports.getEndpointVersionForUlnVersion = getEndpointVersionForUlnVersion;
exports.getNetworkForChainId = getNetworkForChainId;
exports.getNetworksForStage = getNetworksForStage;
exports.networkToChain = networkToChain;
exports.networkToChainType = networkToChainType;
exports.networkToEndpointDomain = networkToEndpointDomain;
exports.networkToEndpointId = networkToEndpointId;
exports.networkToEnv = networkToEnv;
exports.networkToStage = networkToStage;
